### PR TITLE
implement podcast functionality

### LINKED ISSUE #359: Empty UI when playing a podcast episode
**Is your feature request related to a problem? Please describe.**

If I listen to a podcast on my phone, then switch to spotify_player on my laptop, then spotify_player doesn't show anything about the active episode. I can play it or pause it alright (never tried more), but there's nothing displayed about it. This is pretty confusing.

**Describe the solution you'd like**

Basic information about the current episode should be displayed.

**Describe alternatives you've considered**

None.

**Additional context**

There's actually no need for full support for podcasts here (although that could be nice). Just, you know, show what's currently playing :)

<details>
<summary>Screenshot</summary>

Quick example: a podcast is playing (as shown in the Spotify Web UI), the current player is `spotify_player` on my laptop (device name "Magni"), but nothing is visible in the "Playback" section of the TUI:

image

</details>

### LINKED ISSUE #457: Podcast Support
**Is your feature already implemented in the latest `master`?**
No

**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
I would like to have podcast support available in the app. When I search for podcasts, I don't get any of the results I'm expecting. Possibly another pane with podcast results only. I haven't dabbled with Spotifys Podcast API so I'm not sure how the implementation would be done.

### PR DESCRIPTION
**Pull request by SebRollen**
Work-in-progress implementation of podcasts. Still needs quite a bit of work, but this is a large change so I thought I'd open it up in draft mode in case anyone has any comments along the way.

Playback is working for podcast episodes, as is searching for specific episodes. But the playback is still not being populated—in the process of tracking down why that isn't happening now

Fixes #359
Fixes #457

Checklist (based on https://github.com/aome510/spotify-player/issues/457#issuecomment-2156052760)
- [x] Search
  - [x] Shows (podcasts)
  - [x] Episodes (podcast episodes)
- [x] Context menu
  - [x] Shows
    - [x] Copy link
    - [x] Follow
  - [x] Episodes
    - [x] Go to show
    - [x] Copy link
    - [x] Add to playlist
    - [x] Add to queue
- [x] Show detail page
- [x] Playback
  - [x] Play audio
  - [x] Show playback status

## Screenshots

Search page
Screenshot 2024-08-22 at 10 14 06 PM

Show detail page
Screenshot 2024-08-22 at 11 05 13 PM

### CODE REVIEW

File: spotify_player/src/state/model.rs [hunk 1]
```diff
@@ -55,7 +56,42 @@ pub enum ContextId {
...
+        }
+    }
+}
+
+impl Clone for Playback {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Context(context_id, offset) => Self::Context(context_id.clone(), offset.clone()),
+            Self::URIs(playable_ids, offset) => Self::URIs(
+                playable_ids.into_iter().map(|x| x.clone_static()).collect(),
```
Comment (SebRollen) [inline spotify_player/src/state/model.rs:90]: There's a lot of painful places like this throughout the diff where things would be easier if `PlayableId` had `impl Clone`. Will open a PR against `rspotify` to see if that could be added

Comment (SebRollen) [reply spotify_player/src/state/model.rs:90]: https://github.com/ramsayleung/rspotify/pull/490

File: spotify_player/src/client/request.rs [hunk 2]
```diff
@@ -38,6 +38,7 @@ pub enum ClientRequest {
     },
     Search(String),
     AddTrackToQueue(TrackId<'static>),
+    AddEpisodeToQueue(EpisodeId<'static>),
```
Comment (SebRollen) [inline spotify_player/src/client/request.rs:41]: Would love to combine these into a `AddItemToQueue(PlayableId<'static>)`, but can't do that yet as `PlayableId` doesn't `impl Debug` or `Clone`

**Comment by SebRollen** [comment]
> Anything I can do to help? I would love this feature!

The biggest outstanding "unknown" is why the playback screen is still blank after starting playback for an episode, so any help diagnosing this would be fantastic.

The info in https://github.com/aome510/spotify-player/issues/359 may be helpful here. I was hopeful that updating `streaming.rs` to handle episode player events would fix it, but no dice so far.

In general, I'm probably gonna be a bit short on time to be able to push this PR forward this week, but should have more time next week.

**Comment by SebRollen** [comment]
This pull-request is already going to be a doozy to review, so I figured I should open it up for a first look now.

Most of the functionality outlined in #457 is implemented here, save one very important part: rendering the currently playing episode in the Player UI. I've been trying to debug why the playing episode does not show up there, but haven't had too much luck yet. I know that `playback.item` in https://github.com/aome510/spotify-player/pull/552/files#diff-3ee6f331c585285645123ea35a2b8ad7db4d450994a9429b78c4856c6531c410R19 returns `None` even when an episode is playing, but as far as I can tell, `item` is set by Spotify itself, so I'm not sure why it's not `Some(PlayableItem::Episode)`. Would appreciate help in debugging this

**Comment by SebRollen** [comment]
Lots of errors in CI. I've been testing without enabling features—will update the PR to account for the various features as well

**Comment by aome510** [comment]
Thanks @SebRollen for the PR and implementing this big feature! I'm still half-way reviewing it. Just to give you an update. Sorry for the delay cause I haven't found time to work on this project lately.

**Comment by SebRollen** [comment]
@aome510 it's all good, I know it's quite a few changes to review in one go. 

If it's helpful, I'm happy to try to break it into smaller chunks if you're willing to merge in partial updates to the feature. I found it hard initially to find a natural point to split the PR into smaller chunks without merging in a half-baked feature, but maybe we could get all the "backend" parts before adding in the UI updates? Let me know what's most helpful for you

**Review by aome510 — APPROVED** [review]

**Comment by aome510** [comment]
@SebRollen I finally had a chance to get back to this PR and finish it. I've also added playback support. The remaining item is to implement **command handler** for **episode table window**, which I will address in a future PR.

**Pull request merged**